### ISSUE TITLE
header on a title and toc pages

### ISSUE THREAD

**Issue by habamax** [asciidoctor/asciidoctor-pdf#606]
Is there a way in theme customization to have header on a title and toc page?

I have the following in a theme:

```
 header:
   font_size: $base_font_size_small
   height: $base_line_height_length * 6.0
   vertical_align: middle
   recto:
       columns: "<30% 40% >30%"
       left:
         content: 'image:logo/logo-1.png[a, scaledwidth="20%"]'
       right:
         content: 'image:logo/logo-2.png[a, scaledwidth="20%"]'
   verso:
       columns: "<30% 40% >30%"
       left:
         content: 'image:logo/logo-1.png[a, scaledwidth="20%"]'
       right:
         content: 'image:logo/logo-2.png[a, scaledwidth="20%"]'
```

**Comment by pblew** [comment]
I don't believe so currently.  I've resorted to changing the following in `converter.rb` (line 161 in my version):

```ruby
num_front_matter_pages = page_number - 1
```

to

```ruby
num_front_matter_pages = (doc.attr 'pdf-start-running-content-page', page_number).to_i - 1
```

This allows me to put the following into my document to control the page on which running headers and footer start while existing documents get the current behaviour:

```
:pdf-start-running-content-page: 1
```

Hope this helps.

**Comment by deevodavis71** [comment]
I have a very similar requirement to have headers and footers displayed on the Front Page and TOC (it's for the aerospace/military industry and have zero flexibility on this!). 

It would be nice to at least have the option to switch on headers and footers on these pages, even if by default it's switched off for the rest of the community.

**Comment by rockyallen** [comment]
Would be good for me too. Could it also have running headers/footers, roman numeral page numbers for the TOC and preface, and the first chapter starting at page 1?

**Comment by marcdexet-cnrs** [comment]
As @deevodavis71 I have Quality Documentation to generate without flexibility, that's (one of ) the last hurdle to switch people from MS Office to Asciidoc.

**Comment by jhines0042** [comment]
I would like to +1 this request

**Comment by voz** [comment]
+1 for the record. We need it to comply with the legal requirements.

**Comment by themowski** [comment]
Chiming in, someone on my team pointed out that the TOC doesn't have the same header/footer as the rest of the document. I had never noticed it but now that I have, I recognize that it's wrong and needs to be fixed.

I can also see the utility of having the header/footer on the title page. We don't need that now, but might in the future. Allowing a user to specify a specific (possibly different) header/footer for the title page might actually be a more useful feature.

**Comment by danielegalluccio** [comment]
Similar to the above, when writing technical documentation (especially for Regulated environments) it's mandatory to have Header & Footer on _every_ page.
I guess a flag/option to enable/disable it or a way to control it via the custom theme yaml would be a great way to support both use cases

**Comment by habamax** [comment]
I have tried to find where it could be tinkered in the src...

So it is `converter.rb` line 203 (here I have changed it to 0): 

    num_front_matter_pages = 0 #(@index.start_page_number = page_number) - 1

This is the number of skipped pages to generate header and footer later on lines 215-222:

    if page_count > num_front_matter_pages
      unless doc.noheader || @theme.header_height.to_f.zero?
        layout_running_content :header, doc, skip: num_front_matter_pages
      end
      unless doc.nofooter || @theme.footer_height.to_f.zero?
        layout_running_content :footer, doc, skip: num_front_matter_pages
      end
    end

So basically, with that small change I can see header and footer on title and toc pages.

![image](https://user-images.githubusercontent.com/234774/46882186-6f9efd80-ce56-11e8-83ed-02d4ade2991f.png)


The only issue I can immediately see is with page numbering -- it starts from the title page now. Maybe this is not an issue - not sure yet. :)

For sure it is not "production" ready -- it clearly should controlled from theme file -- and I am not sure how it has to be implemented.

**Comment by habamax** [comment]
And example of header theming:

    header:
      pages: all

change in `converter.rb`:

    if page_count > num_front_matter_pages
      unless doc.noheader || @theme.header_height.to_f.zero?
        if (@theme.header_pages == 'all')
          layout_running_content :header, doc, skip: 0
        else
          layout_running_content :header, doc, skip: num_front_matter_pages
        end
      end
      unless doc.nofooter || @theme.footer_height.to_f.zero?
        layout_running_content :footer, doc, skip: num_front_matter_pages
      end
    end

**Comment by habamax** [comment]
@mojavelinux I can try to come up with a PR implementing it, but not sure what kind of stuff should be implemented (and the probable impact on another parts of asciidoctor).

1. Name of the option(s) and values? 
2. Should the page numbering always start after the toc?
3. Should we have option to put header/footer on title+toc, toc without title VS to put it everywhere, everywhere except title and toc.

**Comment by mojavelinux** [comment]
When we started out, we had headers and footers on the toc page. People complained they shouldn't be on that page. So we removed them. Now we have a request to add them back. :scream: So what does that mean? An option, of course!

@habamax Thank you for offering to create PR. Based on your proposal, my advice is to break this down into smaller parts. Trying to do multiple changes at once almost always leads to a stalled change. If we take it one thing at a time, the changes will be much more likely to go through.

Let's first address the issue of which pages the header / footer should be on. This is definitely something that belongs in the theme file. I'd be happy to start with your proposal.

```
header:
  pages: all
```

I think we can assume that "all" does not mean the cover pages.

Another possible option is:

```
header:
  pages: +title,+toc
```

That might be more clear and scalable.

For now, stick with numbering pages as they are. That should be a separate issue.

**Comment by mojavelinux** [comment]
> @themowski wrote: \
> someone on my team pointed out that the TOC doesn't have the same header/footer as the rest of the document. I had never noticed it but now that I have, I recognize that it's wrong and needs to be fixed

you mean the toc page doesn't have a header/footer at all, right?

**Comment by mojavelinux** [comment]
> Please don't implement only half the solution.

I'm not suggesting that we do. But combining two issues into one means you get 0 solutions (which I can say speaking from lots of experience). These changes need to be handled in separate PRs.

**Comment by themowski** [comment]
> 
> 
> > @themowski wrote: 
> > someone on my team pointed out that the TOC doesn't have the same header/footer as the rest of the document. I had never noticed it but now that I have, I recognize that it's wrong and needs to be fixed
> 
> you mean the toc page doesn't have a header/footer at all, right?

Ah, yes, sorry for the misleading wording! You're correct, I mean to say that there is no header/footer on the TOC.

Also -- +1 for adding customization to specify which, if any or both, of the title/toc get headers. I like the proposed `pages: +title,+toc` option for themes. Having the ability to put information (such as PROPRIETARY labels) on the title page in the footer would be a huge benefit, if we can't do that already.

**Comment by nattgris** [comment]
Absolutely, but I'm afraid the issues may not be so easy to separate. For example, it's easy to make a change that appears to work, but where the page number in the header is out of sync with the TOC and/or the outline. The header theming example by @habamax above has that problem.

It may be easier to split the change something like this:

1. Add options to say whether the TOC, title+TOC, or none of them shall be considered part of the front matter. Numbering in this case will, as today, start from the page after the front matter and thus on the first page with a header (since headers are only rendered on non-front matter pages).
2. Add options to specify where numbering shall start, and make sure that headers/footers before that page gets numbered correctly (possibly with roman numerals, as in the outline?). This is probably a more complicated and less well-defined issue.

In any case, when testing, make sure to try with both header and footer containing "content: '{page-number}({page-count})'" and that they get numbers that correspond to both the TOC and the outline.

**Comment by mojavelinux** [comment]
@habamax Are you willing to submit a proposal to implement support for the following?

```
header:
  pages: +title,+toc
```

Just split on comma, then look for `+title` or `+toc`, similar to what your patch was doing when looking for `all`.

**Comment by habamax** [comment]
@mojavelinux sure, I have just to find some time to figure out how to do it with PRs. Last time I did it was years ago :)

**Comment by mojavelinux** [comment]
There's no need to use any PRs. Just upgrade to the latest release of Asciidoctor and you should see that the issue is resolved.

**Comment by habamax** [comment]
@mojavelinux I guess you've mixed our conversations: the one with diagrams and this one :)

**Comment by mojavelinux** [comment]
Ooops! My mistake.

In that case, let me know if you need any assistance getting set up again.

**Comment by habamax** [comment]
Hi @mojavelinux,

1. I wasn't able to implement theming with `+title,+toc`.
It was not clear how to treat only `+title`. Should we have header on title and all except toc pages? If so -- I don't know how to do it :)
2. Instead I have `startfrom: title|toc` option meaning the header(footer) will start either from title or toc pages.

I couldn't find how to fix page numbering issue...

PS, PR stuff is not easy when you have 2 years old fork.

**Comment by habamax** [comment]
It looks like it is better to have this option not per header and footer, but probably more general one, like for a page.

For now page, with my PR, numbering works if both header and footer are started from the same page. Otherwise there is a discrepancy: header and footer shows different page numbers.

**Comment by landshark666** [comment]
@habamax Thank you for implementing this. I had the same issue for the same reason. I've pulled your pull request into my local repo and it has worked great so far.

**Comment by habamax** [comment]
@landshark666, I am thinking of changing the theme params to have only 1 common option both for header and footer, just to avoid page numbering issue. Just not sure how to name it :)

Also @mojavelinux approvement would be nice to have

**Comment by landshark666** [comment]
@habamax I don't have a use for it, but someone might find the ability to start the header/footer on different pages useful. Could a new command such as `:pageNumber: 1` be introduced which explicitly sets the page from the current page forward?

**Comment by habamax** [comment]
@landshark666 I will have a look into it, but I doubt it would be easy to do.

**Comment by Alyaksej** [comment]
In my document, I set 
                                          **start_at:** toc
                                           **footer:**
                                             **...**
                                              **recto:**
                                                 **right:**
                                                   **content:  '*{chapter-title}*'**
I got
![1](https://user-images.githubusercontent.com/36542108/56798679-efebe580-681f-11e9-949f-a6dfea6831d6.png)
How can I change  'Preface' to 'TOC' on footer of TOC or just remove this word from footer?

**Comment by mojavelinux** [comment]
@Alyaksej Please open a separate issue. This one is closed and what you're asking for is an enhancement (albeit a logical one).
